contribs/ev/src/main/java/org/matsim/contrib/ev/stats/EvMobsimListener.java

```diff
@@ -85,4 +100,32 @@ public void notifyMobsimBeforeCleanup(MobsimBeforeCleanupEvent event) {
 			throw new RuntimeException(e);
 		}
 	}
+
+	/** After each chargeTimeStep, the power transfer from each station to the vehicles present at that station (vehiclesAtCharger) is calculated.
+	 * To do so, the energy at the previous chargeTimeStep needs to be known (vehiclesEnergyPreviousTimeStep). The power is then passed in to the
+	 * chargerPowerTimeProfileCollectorProvider. */
+	@Override
+	public void notifyMobsimAfterSimStep(MobsimAfterSimStepEvent event) {
+		if ((event.getSimulationTime() + 1) % evConfigGroup.chargeTimeStep == 0) {
+			ElectricFleet evFleet = chargerPowerTimeProfileCollectorProvider.getEvFleet();
+			Map<Id<Charger>, List<Id<Vehicle>>> vehiclesAtCharger = chargerPowerTimeProfileCollectorProvider.getVehiclesAtCharger();
+			Map<Id<Vehicle>, Double> vehiclesEnergyPreviousTimeStep = chargerPowerTimeProfileCollectorProvider.getVehiclesEnergyPreviousTimeStep();
+			vehiclesAtCharger.forEach((charger, vehicleList) -> {
+				if (!vehicleList.isEmpty()) {
+					double energy = vehicleList.stream().mapToDouble(vehicleId -> EvUnits.J_to_kWh((Objects.requireNonNull(evFleet.getElectricVehicles().get(vehicleId)).getBattery()
+							.getCharge() - vehiclesEnergyPreviousTimeStep.get(vehicleId))*(3600.0/evConfigGroup.chargeTimeStep))).sum();
+					if (!Double.isNaN(energy) && !(energy == 0.0)) {
+						chargerPowerTimeProfileCollectorProvider.setChargerEnergy(charger, energy);
+						vehicleList.forEach(vehicleId -> chargerPowerTimeProfileCollectorProvider
+								.setVehiclesEnergyPreviousTimeStep(vehicleId, Objects.requireNonNull(evFleet.getElectricVehicles().get(vehicleId)).getBattery().getCharge()));
+					} else {
+						chargerPowerTimeProfileCollectorProvider.setChargerEnergy(charger, 0.0);
+					}
+				} else {
+					chargerPowerTimeProfileCollectorProvider.setChargerEnergy(charger, 0.0);
+				}
+			});
+
+		}
+	}
```

Comment:
Please move this method to `ChargerPowerTimeProfileCollectorProvider`, so all logic is in one place.

Reply:
I would love to, but I cannot move the `MobsimAfterSimStepListener `to `ChargerPowerTimeProfileCollectorProvider` in a way that actually works. By just moving it, `notifyMobsimAfterSimStep` is never called. I guess there is something wrong with the binding of the class. Do you have any idea of what could be the cause?

Reply:
A few things will need to be rearranged here to make it work. I will try to look into it next week and suggest some fixes.